### PR TITLE
add bocc keyboard

### PR DESCRIPTION
**Pull request by moyi4681**

### CODE REVIEW

File: keyboards/dztech/bocc/config.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x445A  // "DZ"
+#define PRODUCT_ID      0x1010
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    BOCC
+#define PRODUCT         BOCC
+#define DESCRIPTION     BOCC Keyboard
```
Comment (ridingqwerty) [inline keyboards/dztech/bocc/config.h:11]:
```suggestion
```

File: keyboards/dztech/bocc/rules.mk [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes           # USB Nkey Rollover
+BACKLIGHT_ENABLE = yes      # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
```
Comment (ridingqwerty) [inline keyboards/dztech/bocc/rules.mk:31]:
```suggestion
```

File: keyboards/dztech/bocc/bocc.c [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#include "bocc.h"
```
Comment (drashna) [inline keyboards/dztech/bocc/bocc.c:1]: Could you add a license header?

File: keyboards/dztech/bocc/bocc.h [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#pragma once
```
Comment (drashna) [inline keyboards/dztech/bocc/bocc.h:1]: Could you add a license header?

File: keyboards/dztech/bocc/config.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
+#pragma once
```
Comment (drashna) [inline keyboards/dztech/bocc/config.h:1]: Could you add a license header?

File: keyboards/dztech/bocc/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,33 @@
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/dztech/bocc/keymaps/default/keymap.c:1]: Could you add a license header?

File: keyboards/dztech/bocc/bocc.c [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+void matrix_init_kb(void) {
```
Comment (fauxpark) [inline keyboards/dztech/bocc/bocc.c:2]:
```suggestion

void matrix_init_kb(void) {
```

File: keyboards/dztech/bocc/bocc.c [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+    setPinOutput(E6);
+    matrix_init_user();
+}
+bool led_update_kb(led_t led_state) {
```
Comment (fauxpark) [inline keyboards/dztech/bocc/bocc.c:6]:
```suggestion

bool led_update_kb(led_t led_state) {
```

File: keyboards/dztech/bocc/readme.md [hunk 9]
```diff
@@ -0,0 +1,11 @@
+#BOCC
+* Keyboard Maintainer: DZTECH
```
Comment (fauxpark) [inline keyboards/dztech/bocc/readme.md:2]:
```suggestion

* Keyboard Maintainer: DZTECH
```

File: keyboards/dztech/bocc/readme.md [hunk 10]
```diff
@@ -0,0 +1,11 @@
+#BOCC
```
Comment (fauxpark) [inline keyboards/dztech/bocc/readme.md:1]:
```suggestion
# BOCC
```

File: keyboards/dztech/bocc/rules.mk [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+MCU = atmega32u4
+
+# Bootloader selection
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
```
Comment (fauxpark) [inline keyboards/dztech/bocc/rules.mk:11]:
```suggestion
```

File: keyboards/dztech/bocc/rules.mk [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes           # USB Nkey Rollover
+BACKLIGHT_ENABLE = yes      # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
```
Comment (fauxpark) [inline keyboards/dztech/bocc/rules.mk:29]:
```suggestion
BLUETOOTH_ENABLE = no       # Enable Bluetooth
```

File: keyboards/dztech/bocc/config.h [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+ */
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x445A  // "DZ"
+#define PRODUCT_ID      0x1010
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    BOCC
```
Comment (fauxpark) [inline keyboards/dztech/bocc/config.h:24]:
```suggestion
#define MANUFACTURER    DZtech
```
?

File: keyboards/dztech/bocc/keymaps/default/keymap.c [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, 
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO),
+
+	LAYOUT(
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, 
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, 
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, 
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, 
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO),
+
```
Comment (fauxpark) [inline keyboards/dztech/bocc/keymaps/default/keymap.c:47]:
```suggestion
```

File: keyboards/dztech/bocc/keymaps/default/keymap.c [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+	LAYOUT(
+		KC_F1, KC_F2, KC_ESC, KC_1, KC_2, KC_3, KC_4, KC_5, KC_6, KC_7, KC_8, KC_9, KC_0, KC_MINS, KC_EQL,  KC_GRV,  KC_BSPC, KC_HOME, 
+		KC_F3, KC_F4, KC_TAB, KC_Q, KC_W, KC_E, KC_R, KC_T, KC_Y, KC_U, KC_I, KC_O, KC_P, KC_LBRC, KC_RBRC, KC_BSLS, KC_PGUP, 
+		KC_F5, KC_F6, KC_CAPS, KC_A, KC_S, KC_D, KC_F, KC_G, KC_H, KC_J, KC_K, KC_L, KC_SCLN, KC_QUOT, KC_ENT, KC_PGDN, 
+		KC_F7, KC_F8, KC_LSFT, KC_Z, KC_X, KC_C, KC_V, KC_B, KC_N, KC_M, KC_COMM, KC_DOT, KC_SLSH, KC_RSFT, KC_UP, KC_END, 
+		KC_F9, KC_F10, KC_LCTL, KC_LGUI, KC_LALT, KC_BSLS, KC_SPC, KC_RALT, KC_RCTL, KC_LEFT,  KC_DOWN, KC_RGHT),
```
Comment (noroadsleft) [inline keyboards/dztech/bocc/keymaps/default/keymap.c:25]: Where, geographically speaking, is the `KC_BSLS` that is present on this line?

image

File: keyboards/dztech/bocc/bocc.h [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include "quantum.h"
+
+#define LAYOUT( \
```
Comment (noroadsleft) [inline keyboards/dztech/bocc/bocc.h:20]: As it seems this PCB supports more than one layout...

```suggestion
#define LAYOUT_all( \
```

File: keyboards/dztech/bocc/info.json [hunk 17]
```diff
@@ -0,0 +1,12 @@
+{
+    "keyboard_name": "BOCC", 
+    "url": "", 
+    "maintainer": "DZTECH", 
+    "width": 18.5, 
+    "height": 5, 
+    "layouts": {
+        "LAYOUT": {
```
Comment (noroadsleft) [inline keyboards/dztech/bocc/info.json:8]:
```suggestion
        "LAYOUT_all": {
```

File: keyboards/dztech/bocc/keymaps/default/keymap.c [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+	LAYOUT(
```
Comment (noroadsleft) [inline keyboards/dztech/bocc/keymaps/default/keymap.c:20]:
```suggestion
	LAYOUT_all(
```

File: keyboards/dztech/bocc/keymaps/via/keymap.c [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+	LAYOUT(
```
Comment (noroadsleft) [inline keyboards/dztech/bocc/keymaps/via/keymap.c:20]:
```suggestion
	LAYOUT_all(
```

File: keyboards/dztech/bocc/keymaps/via/keymap.c [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+		KC_F1, KC_F2, KC_ESC, KC_1, KC_2, KC_3, KC_4, KC_5, KC_6, KC_7, KC_8, KC_9, KC_0, KC_MINS, KC_EQL,  KC_GRV,  KC_BSPC, KC_HOME, 
+		KC_F3, KC_F4, KC_TAB, KC_Q, KC_W, KC_E, KC_R, KC_T, KC_Y, KC_U, KC_I, KC_O, KC_P, KC_LBRC, KC_RBRC, KC_BSLS, KC_PGUP, 
+		KC_F5, KC_F6, KC_CAPS, KC_A, KC_S, KC_D, KC_F, KC_G, KC_H, KC_J, KC_K, KC_L, KC_SCLN, KC_QUOT, KC_ENT, KC_PGDN, 
+		KC_F7, KC_F8, KC_LSFT, KC_Z, KC_X, KC_C, KC_V, KC_B, KC_N, KC_M, KC_COMM, KC_DOT, KC_SLSH, KC_RSFT, KC_UP, KC_END, 
+		KC_F9, KC_F10, KC_LCTL, KC_LGUI, KC_LALT, KC_BSLS, KC_SPC, KC_RALT, KC_RCTL, KC_LEFT,  KC_DOWN, KC_RGHT),
+
+	LAYOUT(
```
Comment (noroadsleft) [inline keyboards/dztech/bocc/keymaps/via/keymap.c:27]:
```suggestion
	LAYOUT_all(
```

File: keyboards/dztech/bocc/keymaps/via/keymap.c [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, 
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, 
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, 
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, 
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO),
+
+	LAYOUT(
```
Comment (noroadsleft) [inline keyboards/dztech/bocc/keymaps/via/keymap.c:34]:
```suggestion
	LAYOUT_all(
```

File: keyboards/dztech/bocc/keymaps/via/keymap.c [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, 
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, 
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, 
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, 
+		KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO),
+
+	LAYOUT(
```
Comment (noroadsleft) [inline keyboards/dztech/bocc/keymaps/via/keymap.c:41]:
```suggestion
	LAYOUT_all(
```

File: keyboards/dztech/bocc/bocc.h [hunk 23]
```diff
@@ -21,8 +21,8 @@
 	K000, K001, K002, K003, K004, K005, K006, K007, K008, K009, K010, K011, K012, K013, K014, K414, K015, K016, \
 	K100, K101, K102, K103, K104, K105, K106, K107, K108, K109, K110, K111, K112, K113, K114, K115, K116,       \
 	K200, K201, K202, K203, K204, K205, K206, K207, K208, K209, K210, K211, K212, K213, K214,       K216,       \
-	K300, K301, K302, K303, K304, K305, K306, K307, K308, K309, K310, K311, K312, K313,       K315, K316,       \
-	K400, K401, K402, K403, K404, K405,       K407,                   K411, K412, K413,       K415, K416        \
+	K300, K405, K301, K302, K303, K304, K305, K306, K307, K308, K309, K310, K311, K312, K313, K315, K316,       \
```
Comment (noroadsleft) [inline keyboards/dztech/bocc/bocc.h:24]: You moved the keycode in the keymaps to the fourth position on this row, but the identifier in the macro to the second.

```suggestion
	K300, K301, K302, K405, K303, K304, K305, K306, K307, K308, K309, K310, K311, K312, K313, K315, K316,       \
```

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**